extension/chrome/elements/compose-modules/compose-send-btn-module.ts

```diff
@@ -194,65 +214,62 @@ export class ComposeSendBtnModule extends ViewModule<ComposeView> {
     // if this is a password-encrypted message, then we've already shown progress for uploading to backend
     // and this requests represents second half of uploadable effort. Else this represents all (no previous heavy requests)
     const progressRepresents = this.view.pwdOrPubkeyContainerModule.isVisible() ? 'SECOND-HALF' : 'EVERYTHING';
-    let msgSentRes: GmailRes.GmailMsgSend;
-    try {
-      this.isSendMessageInProgress = true;
-      msgSentRes = await this.view.emailProvider.msgSend(msg, (p) => this.renderUploadProgress(p, progressRepresents));
-    } catch (e) {
-      if (msg.thread && ApiErr.isNotFound(e) && this.view.threadId) { // cannot send msg because threadId not found - eg user since deleted it
-        msg.thread = undefined;
-        msgSentRes = await this.view.emailProvider.msgSend(msg, (p) => this.renderUploadProgress(p, progressRepresents));
-      } else {
-        this.isSendMessageInProgress = false;
-        throw e;
-      }
-    }
-    BrowserMsg.send.notificationShow(this.view.parentTabId, { notification: `Your ${this.view.isReplyBox ? 'reply' : 'message'} has been sent.` });
-    BrowserMsg.send.focusBody(this.view.parentTabId); // Bring focus back to body so Gmails shortcuts will work
-    const operations = [this.view.draftModule.draftDelete()];
-    if (msg.externalId) {
-      operations.push((async (externalId, id) => {
-        const gmailMsg = await this.view.emailProvider.msgGet(id, 'metadata');
-        const messageId = GmailParser.findHeader(gmailMsg, 'message-id');
-        if (messageId) {
-          await this.view.acctServer.messageGatewayUpdate(externalId, messageId);
-        } else {
-          Catch.report('Failed to extract Message-ID of sent message');
+    const sentIds: string[] = [];
+    const supplementaryOperations: Promise<void>[] = [];
+    const supplementaryOperationsErrors: any[] = []; // tslint:disable-line:no-unsafe-any
+    const success: EmailParts[] = [];
+    const failures: { recipient: EmailParts, e: any }[] = [];
+    for (const msg of msgObj.msgs) {
+      let msgSentRes: GmailRes.GmailMsgSend;
+      const msgRecipients = msg.getAllRecipients();
+      while (true) { // first try with msg.thread, and then possibly try again without it
+        try {
+          msgSentRes = await this.view.emailProvider.msgSend(msg, (p) => this.renderUploadProgress(p, progressRepresents));
+          success.push(...msgRecipients);
+          sentIds.push(msgSentRes.id);
+          if (msg.externalId) {
+            const operation = (async (externalId, id) => {
+              const gmailMsg = await this.view.emailProvider.msgGet(id, 'metadata');
+              const messageId = GmailParser.findHeader(gmailMsg, 'message-id');
+              if (messageId) {
+                await this.view.acctServer.messageGatewayUpdate(externalId, messageId);
+              } else {
+                throw new Error('Failed to extract Message-ID of sent message');
+              }
```

Comment:
The contents of `operation` could be their own, appropriately named method. Leading to a bit less clutter here.